CFE-4099: Fixed bug where order of keys breaks set-input

larsewi opened the pull request:
Fix bug where order of keys breaks set-input

Fixed bug where changing the order of keys when using `cfbs set-input`
causes it to error with the following error message `ERROR: Input data
for module 'delete-files' does not conform with input definition`

Ticket: CFE-4099
Changelog: None
Signed-off-by: Lars Erik Wik <lars.erik.wik@northern.tech>

larsewi committed:
Fixed bug in shell test

Fixed issue where shell test 031_get_set_input_pipe.sh was deleting
wrong module directory.

Ticket: None
Changelog: None
Signed-off-by: Lars Erik Wik <lars.erik.wik@northern.tech>

larsewi committed:
Fixed bug where order of keys breaks set-input

Fixed bug where changing the order of keys when using `cfbs set-input`
causes it to error with the following error message `ERROR: Input data
for module 'delete-files' does not conform with input definition`

Before two `dict`'s of subclass `OrderedDict` got compared two each
other using the `_compare_dict` function. Then they were later compared
using the `==` operator. Due to polymorphism the `==` operator would
also compare the order of the keys, thus return `False`.

Ticket: CFE-4099
Changelog: None
Signed-off-by: Lars Erik Wik <lars.erik.wik@northern.tech>

olehermanse reviewed (approved):
LGTM, thanks :)

vpodzime commented on cfbs/commands.py:
```diff
@@ -1039,12 +1039,13 @@ def _compare_list(a, b):
+            if isinstance(x, dict) and _compare_dict(x, y):
+                continue
+            if isinstance(x, list) and _compare_list(x, y):
+                continue
```
Are you sure this is the right thing to do? Please extend the commit message with an explanation of how exactly this change fixes the issue.

olehermanse replied:
Talked to Lars in person. Although the current approach works, it's probably safer (and more correct) to just do a choice between the 3 options (dict, list, other) and then do the correct comparison inside that block. The current approach with `and` and this kind of "fallthrough" to the logic below, is confusing.

Specifically, it is a bit weird to think about what happens when it is a dict, but `_compare_dict()` returns False, causing us to go down to comparing them again.

vpodzime reviewed (commented):
So this resolves https://github.com/cfengine/cfbs/pull/139#discussion_r1009534663, thanks. Can you also please fix https://github.com/cfengine/cfbs/pull/142/files#r1026193657 here?

larsewi committed:
Fixed bug in default parameter of `_compare_dict`

Fixed bug where modifications of default parameter `ignore` in
function `_compare_dict` is kept between function calls.

Ticket: None
Changelog: None
Signed-off-by: Lars Erik Wik <lars.erik.wik@northern.tech>

olehermanse reviewed (approved):
Thanks!

vpodzime commented on cfbs/commands.py:
```diff
@@ -1023,9 +1023,10 @@ def set_input_command(name, infile):
+        assert ignore is None or isinstance(ignore, set)
         assert isinstance(a, dict) and isinstance(b, dict)
-        if set(a.keys()) != set(b.keys()) - ignore:
+        if set(a.keys()) != set(b.keys()) - (set() if ignore is None else ignore):
```
The common trick is to do `ignore = ignore or set()` (or `ignore = ignore if ignore is not None else set()`) at the beginning of the function.

olehermanse commented on tests/shell/032_get_set_input_unordered.sh:
```diff
@@ -0,0 +1,42 @@
+    "variable": "files",
+    "while": "Specify another file you want deleted on your hosts?"
+  }
+]' | cfbs --log=debug set-input delete-files -
```
To me it's nicer / easier the way he wrote it
